app/src/behaviors/behavior_hold_tap.c

```diff
@@ -0,0 +1,491 @@
+		}
+		ZMK_EVENT_RELEASE_AGAIN(captured_event);
+	}
+}
+
+
+/************************************************************    ACTIVE TAP HOLD HELPER FUNCTIONS */
+
+static struct active_hold_tap *find_hold_tap(u32_t position)
+{
+	for (int i = 0; i < ZMK_BHV_HOLD_TAP_MAX_HELD; i++) {
+		if (active_hold_taps[i].position == position) {
+			return &active_hold_taps[i];
+		}
+	}
+	return NULL;
+}
+
+static struct active_hold_tap *store_hold_tap(u32_t position, u32_t param_hold, u32_t param_tap, const struct behavior_hold_tap_config *config)
+{
+	for (int i = 0; i < ZMK_BHV_HOLD_TAP_MAX_HELD; i++) {
+		if (active_hold_taps[i].position != ZMK_BHV_HOLD_TAP_POSITION_NOT_USED) {
+			continue;
+		}
+		active_hold_taps[i].position = position;
+		active_hold_taps[i].is_decided = false;
+		active_hold_taps[i].is_hold = false;
+		active_hold_taps[i].config = config;
+		active_hold_taps[i].param_hold = param_hold;
+		active_hold_taps[i].param_tap = param_tap;
+		return &active_hold_taps[i];
+	}
+	return NULL;
+}
+
+static void clear_hold_tap(struct active_hold_tap *hold_tap)
+{
+	hold_tap->position = ZMK_BHV_HOLD_TAP_POSITION_NOT_USED;
+	hold_tap->is_decided = false;
+	hold_tap->is_hold = false;
+	hold_tap->work_is_cancelled = false;
+}
+
+enum decision_moment {
+	HT_KEY_UP               = 0,
+	HT_OTHER_KEY_DOWN       = 1,
+	HT_OTHER_KEY_UP         = 2,
+	HT_TIMER_EVENT          = 3,
+};
+
+static void decide_balanced(struct active_hold_tap *hold_tap, enum decision_moment event)
+{
+	switch (event) {
+	case HT_KEY_UP:
+		hold_tap->is_hold = 0;
+		hold_tap->is_decided = true;
+		break;
+	case HT_OTHER_KEY_UP:
+		hold_tap->is_hold = 1;
+		hold_tap->is_decided = true;
+		break;
+	case HT_TIMER_EVENT:
+		hold_tap->is_hold = 1;
+		hold_tap->is_decided = true;
+		break;
+	default: return;
+	}
+}
+
+static void decide_tap_preferred(struct active_hold_tap *hold_tap, enum decision_moment event)
+{
+	switch (event) {
+	case HT_KEY_UP:
+		hold_tap->is_hold = 0;
+		hold_tap->is_decided = true;
+		break;
+	case HT_TIMER_EVENT:
+		hold_tap->is_hold = 1;
+		hold_tap->is_decided = true;
+		break;
+	default: return;
+	}
+}
+
+static void decide_hold_preferred(struct active_hold_tap *hold_tap, enum decision_moment event)
+{
+	switch (event) {
+	case HT_KEY_UP:
+		hold_tap->is_hold = 0;
+		hold_tap->is_decided = true;
+		break;
+	case HT_OTHER_KEY_DOWN:
+		hold_tap->is_hold = 1;
+		hold_tap->is_decided = true;
+		break;
+	case HT_TIMER_EVENT:
+		hold_tap->is_hold = 1;
+		hold_tap->is_decided = true;
+		break;
+	default: return;
+	}
+}
+
+static void decide_hold_tap(struct active_hold_tap *hold_tap, enum decision_moment event)
+{
+	if (hold_tap->is_decided) {
+		return;
+	}
+
+	if (hold_tap != undecided_hold_tap) {
+		LOG_DBG("ERROR found undecided tap hold that is not the active tap hold");
+		return;
+	}
+
+	char *flavor = hold_tap->config->flavor;
+	if (strcmp(flavor, "balanced") == 0) {
```

Comment:
We shouldn't be doing string comparisons for this. It's inefficient, and Zephyr offers a better option here. Zephyr will assign numeric values for each enum entry, and those values are what we should be storing, and comparing, when making these "which flavor is it" checks.

Reply:
done. i couldn't figure out how to get the string value at runtime, so there's a rather ugly log_dbg statement now.